File: api/src/models/festival.js
// models/festival.model.js
const mongoose = require("mongoose");

const festivalSchema = new mongoose.Schema(
  {
    title: {
      type: String,
      required: [true, "Title cannot be blank"],
      trim: true,
      maxlength: 255,
    },
    image: {
      type: String,
      required: [true, "Image is required"],
      default: "default.jpg",
    },
    location: {
      type: String,
      trim: true,
      maxlength: 255,
    },
    description: {
      type: String,
      trim: true,
    },
    start_time: {
      type: Date,
      required: [true, "Start time is required"],
      validate: {
        validator: function (value) {
          return value >= new Date();
        },
        message: "The start time must be from now on.",
      },
    },
    end_time: {
      type: Date,
      required: [true, "End time is required"],
    },
  },
  {
    timestamps: { createdAt: "created_at", updatedAt: "updated_at" },
  }
);

// Custom validation to ensure end_time is not before start_time
festivalSchema.pre("save", function (next) {
  if (this.end_time && this.start_time && this.end_time < this.start_time) {
    const error = new Error(
      "The end time must not be less than the start time."
    );
    return next(error);
  }
  next();
});

festivalSchema.pre("findOneAndUpdate", function (next) {
  const update = this.getUpdate();
  if (
    update.end_time &&
    update.start_time &&
    update.end_time < update.start_time
  ) {
    const error = new Error(
      "The end time must not be less than the start time."
    );
    return next(error);
  }
  next();
});

module.exports = mongoose.model("Festival", festivalSchema);

File: src/main/java/com/ra/base_spring_boot/model/Festival.java
package com.ra.base_spring_boot.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.ra.base_spring_boot.model.base.BaseObject;
import jakarta.persistence.*;
import jakarta.validation.constraints.FutureOrPresent;
import jakarta.validation.constraints.NotBlank;
import lombok.*;

import java.util.Date;

@Entity
@Table(name = "festivals")
@NoArgsConstructor
@AllArgsConstructor
@Getter
@Setter
@Builder
public class Festival extends BaseObject {

    @NotBlank(message = "Title cannot be blank")
    @Column(nullable = false)
    private String title;

    @Column( nullable = false)
    private String image;

    @Column(name = "location")
    private String location;

    @Column(name = "description", columnDefinition = "TEXT")
    private String description;

    @FutureOrPresent(message = "The start time must be from now on.")
    @Column(name = "start_time", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date startTime;

    @Column(name = "end_time", nullable = false)
    @JsonFormat(pattern = "yyyy-MM-dd'T'HH:mm:ss.SSSXXX", timezone = "Asia/Ho_Chi_Minh")
    private Date endTime;

    @PrePersist
    @PreUpdate
    private void validateDates() {
        if (endTime != null && startTime != null && endTime.before(startTime)) {
            throw new IllegalArgumentException("The end time must not be less than the start time.");
        }
    }
}